### PR TITLE
NRF51822 BLE radio driver (blemin)

### PR DESCRIPTION
**Pull request by riotrf**
Implementation of a Bluetooth Low Energy (BLE) radio driver for the NRF51822!

The following features are not completed yet / development in progress :
- receiving and parsing of BLE incoming data
- establishing BLE Data Channel connections (CONNECT_REQ)
- transmitting and receiving of BLE data packets
- GATT (to heavy, searching for a lightweight alternative...)

ps. big thx to Hauke for his work on nrfmin driver - this driver is based on his initial work.

### CODE REVIEW

File: sys/include/net/ng_ble.h [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+#include "net/ng_netdev.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief   Default channel set after initialization
+ */
+#define BLEMIN_DEFAULT_CHANNEL	37	/* 2402MHz */
```
Comment (d00616) [inline sys/include/net/ng_ble.h]: The comment is not correct.

**Comment by d00616** [comment]
Great work.  At the moment i resear a lot in datasheets. I think the "Programmable Peripheral Interconnect (PPI)" can help to automate some tasks without CPU interaction. I have created an PR to switch periph_rtt from RTC0 to RTC1 so predifined PPI channels can be used to do time depending tasks.

**Comment by riotrf** [comment]
I was just working on my mail backlog and found a mail regarding "Google Summer of Code 2015" - it seems that a Student called James Hollister is currently working on "Bluetooth Low Energy Support" - anyone knows how to reach him?

@d00616 - I am currently working to get the recv part done - and implement a kind of "blemon" shell command that simply monitors BLE incoming packets on a specified channel. I was very happy to see that the datawhitening and the crc calculation is performed "onboard" the nrf. I am still not 100% used to the nrf and still have some optimizations to implement - but I am not fully understanding them etc.
fe.
(opt 1)
/* Lock interframe spacing, so that the radio won't send too soon / start RX too early */
	NRF_RADIO->TIFS = 145;
(opt 2)
/* Copy the BLE override registers from FICR */
RF_RADIO->OVERRIDE0-4
(opt 3)
NRF_RADIO->SHORTS < is that what u mean?

**Comment by OlegHahm** [comment]
> I was just working on my mail backlog and found a mail regarding "Google Summer of Code 2015" - it seems that a Student called James Hollister is currently working on "Bluetooth Low Energy Support" - anyone knows how to reach him?

Try @jhollister ;)

**Comment by jhollister** [comment]
Here! I think we should be able to work together on this (if you want). 
@PeterKietzmann what are your thoughts on how this will affect my project this summer?

**Comment by jhollister** [comment]
I get a long loop of 
```
2015-05-24 15:42:02,753 - INFO # blemin: Error allocating packet payload on RX
```
when I send a packet. Is this just because receiving hasn't been implemented yet or is something wrong with my setup?

**Comment by PeterKietzmann** [comment]
@katezilla could you test if you experience the same error?

**Comment by riotrf** [comment]
BLE development wiki,  https://github.com/riotrf/RIOT.RF/wiki

**Comment by riotrf** [comment]
Tested the ng_sniffer application with latest wireshark 1.12.x version that has BLE dissector/parser included (DLT_BLUETOOTH_LE_LL) - works fine.

**Comment by rfswarm** [comment]
This latest patch enables the BLE Advertising Linkstate (to enable the receive/dump functionality change "_ll_state" to anything but LL_ADVERTISING).

The interface continuously sends BLE Advertising packets (delayed by _advertising_interval) while waiting to receive SCAN_REQ and CONNECTION_REQ packets. If a SCAN_REQ is received the interface will reply with an SCAN_RSP packet while continuing the advertising.

- random 1-10ms BLE Advertising Delay has not yet been implemented
- channel switching has not yet been implemented
- handling of CONNECTION_REQ has not yet been implemented
- BLE data connections at all have not yet been implemented

ps. This patch should make the device "visible" to a BLE scanning device f.e. your mobile phone (so far only tested with Blackberry)

**Comment by riotrf** [comment]
The latest patch enables the BLE Scanning and (the preperation of) BLE Initiating Link state (to enable use corresponding _ll_state value).

I added some kind of packet backlog/cache feature to limit 'endless SCAN_REQ and SCAN_RSP packets. This feature may be removed in following releases - the ble spec is prity frustrating on that point ("backoffCount"..."upperLimit" still not clear to me - and not defined for SCAN_RSPs in ADV ll state hm...) - u can change the current backlog size via (BLE_MAX_CACHE_COUNT) - this is poc code.

The next big task is: Get a BLE data connection working

I will add a shell cmd line 'set arg' to switch BLE link state (for testing purposes) "on the fly" in the next release.

**Comment by PeterKietzmann** [comment]
@riotrf, @rfswarm. Now James opened his PR #3541. Some code in that PR is based on your implementation. Thanks for all your work! I'd like to close your PR for now and hope to get your know-how there. Are you fine with that? To improve the BLE stack in future, you can maybe do small PRs against or based on James PR. Allright?

**Comment by PeterKietzmann** [comment]
@jhollister, is there still functionality inside this PR that is not covered by yours? @riotrf, @haukepetersen, @OlegHahm do you think we can close this PR because of #3541?

**Comment by OlegHahm** [comment]
If all relevant functionality is covered in #3541, too, I don't see a reason against this?

**Comment by jhollister** [comment]
Currently, #3541 does not have scanning and advertising functionality.

**Comment by OlegHahm** [comment]
Anyone planning to work on this?

**Comment by PeterKietzmann** [comment]
#3541 was the more up-to-date PR regarding BLE. As this PR implemented one feature that  #3541 didn't include, I'll close it with memo label set.

**Pull request closed without merging**